runtime-image/test/test.ts

```diff
@@ -139,55 +150,86 @@ const CONFIGURATIONS: TestConfig[] = [
   }
 ];
 
-CONFIGURATIONS.forEach(config => {
-  describe(`nodejs-docker: Image ${config.tag}`, () => {
-    const containerName = uuid.v4();
-    it(config.description, function(done) {
-      this.timeout(2 * TIMEOUT);
-      runDocker(config.tag, containerName, PORT, host => {
-        // Wait for the docker container to start
+describe('runtime image', () => {
+  before(async function() {
+    this.timeout(BUILD_TIMEOUT);
+    await buildDocker(ROOT, 'test/nodejs');
+    const dir = path.join(ROOT, 'test', 'definitions', 'base-install-node');
+    await buildDocker(dir, 'test/definitions/base-install-node');
+  });
+
+  CONFIGURATIONS.forEach(config => {
+    describe(`Image ${config.tag}`, () => {
+      let container: Docker.Container|undefined = undefined;
+      before(async function() {
+        this.timeout(BUILD_TIMEOUT);
+        // converts a/b/c to a/b/c on Unix and a\b\c on Windows
+        const sysDepPath = path.join.apply(null, config.tag.split('/'));
+        const dir = path.join(ROOT, sysDepPath);
+        await buildDocker(dir, config.tag);
+        container = await runDocker(config.tag, PORT);
+      });
+
+      it(config.description, function(done) {
+        this.timeout(2 * RUN_TIMEOUT);
         setTimeout(() => {
           request(`http://${host}:${PORT}`, (err, _, body) => {
             assert.ifError(err);
             assert.equal(body, config.expectedOutput);
             done();
           });
-        }, TIMEOUT);
+        }, RUN_TIMEOUT);
       });
-    });
 
-    after(done => {
-      exec(`docker stop ${containerName}`, (err, stdout, stderr) => {
-        log(stdout);
-        log(stderr);
-        assert.ifError(err);
-        done();
+      after(async () => {
+        if (container) {
+          await container.stop();
+          await container.remove();
+        }
       });
     });
   });
 });
 
-/**
- * Start a docker process for the given test
- */
-function runDocker(tag: string, name: string, port: number,
-                   callback: (host: string) => void) {
-  let d = spawn(
-      'docker',
-      [ 'run', '--rm', '-i', '--name', name, '-p', `${port}:${port}`, tag ]);
-
-  d.stdout.on('data', log);
-  d.stderr.on('data', log);
-  d.on('error', (err) => {
-    log(`Error spawning docker process: ${util.inspect(err)}`);
-    assert.ifError(err);
+function buildDocker(dir: string, tag: string): Promise<any> {
+  const tarStream = tar.pack(dir);
+  return new Promise<any>((resolve, reject) => {
+    DOCKER.buildImage(tarStream, {t : tag}, (err1, stream) => {
+      if (err1) {
+        return reject(err1);
+      }
+
+      function onFinished(err2: Error, output: any) {
+        if (err2) {
+          return reject(err2);
+        }
+
+        resolve(output);
+      }
+
+      function onProgress(event: any) {
+        log(event.stream);
```

Comment:
Curious. Aren't these logged just as `Object [object]` by the `log()` function?

Reply:
`event.stream` etc. are all strings.